File: analysis/model/src/test/java/de/maibornwolff/codecharta/translator/MetricNameTranslatorTest.java
package de.maibornwolff.codecharta.translator;

import com.google.common.collect.ImmutableMap;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class MetricNameTranslatorTest {
    @Test
    public void should_replace_nothing() throws Exception {
        String original = "testestest";
        MetricNameTranslator replacer = MetricNameTranslator.TRIVIAL;

        assertThat(replacer.translate(original), is(original));
    }

    @Test
    public void should_replace_specified_string() throws Exception {
        String original = "oldValue";
        String replacement = "newString";
        MetricNameTranslator replacer = new MetricNameTranslator(ImmutableMap.of(original, replacement));

        assertThat(replacer.translate(original), is(replacement));
    }

    @Test
    public void should_replace_other_string_with_prefixed_underscored_lowercase() throws Exception {
        String original = "some String";
        String expected = "pre_some_string";
        MetricNameTranslator replacer = new MetricNameTranslator(ImmutableMap.of("oldValue", "newValue"), "pre_");

        assertThat(replacer.translate(original), is(expected));
    }

    @Test
    public void should_replaceMany() throws Exception {
        String[] original = new String[]{"this", "that", "other"};
        String[] expected = new String[]{"oooo", "iiii", "other"};
        MetricNameTranslator replacer = new MetricNameTranslator(ImmutableMap.of("this", "oooo", "that", "iiii", "bla", "blubb"));

        assertThat(replacer.translate(original), is(expected));
    }

    @Test(expected = IllegalArgumentException.class)
    public void should_not_validate_with_same_values() {
        new MetricNameTranslator(ImmutableMap.of("this", "that", "these", "that"));
    }

    @Test
    public void should_validate_with_multiple_empty_values() {
        MetricNameTranslator translator = new MetricNameTranslator(ImmutableMap.of("this", "", "these", ""));

        assertThat(translator.translate("this"), is(""));
        assertThat(translator.translate("these"), is(""));
    }
}

File: analysis/model/src/test/kotlin/de/maibornwolff/codecharta/translator/MetricNameTranslatorTest.kt
package de.maibornwolff.codecharta.translator

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import kotlin.test.assertFailsWith

class MetricNameTranslatorTest {
    @Test
    fun `trivial replacer should not replace anything`() {
        val replacer = MetricNameTranslator.TRIVIAL
        val original = "testestest"
        assertThat(replacer.translate(original)).isEqualTo(original)
    }

    @Test
    fun `replacer without specified prefix should replace specified string`() {
        val original = "oldValue"
        val replacement = "newString"
        val replacer = MetricNameTranslator(mapOf(Pair(original, replacement)))
        assertThat(replacer.translate(original)).isEqualTo(replacement)
    }

    @Test
    fun `replacer without specified prefix should not replace other string`() {
        val original = "oldValue"
        val replacement = "newString"
        val replacer = MetricNameTranslator(mapOf(Pair(original, replacement)))
        val otherString = "testestest"
        assertThat(replacer.translate(otherString)).isEqualTo(otherString)
    }

    @Test
    fun `replacer with prefix should replace specified string`() {
        val original = "oldValue"
        val replacement = "newString"
        val replacer = MetricNameTranslator(mapOf(Pair(original, replacement)), "pre_")
        assertThat(replacer.translate(original)).isEqualTo(replacement)
    }

    @Test
    fun `replacer with prefix should prefix other string with underscores`() {
        val original = "oldValue"
        val replacement = "newString"
        val replacer = MetricNameTranslator(mapOf(Pair(original, replacement)), "pre_")
        assertThat(replacer.translate("some String")).isEqualTo("pre_some_string")
    }

    @Test
    fun `replacer should replaceMany`() {
        val replacer =
            MetricNameTranslator(
                mapOf(
                    Pair("this", "oooo"),
                    Pair("that", "iiii"),
                    Pair("bla", "blubb")
                )
            )

        val original: Array<String?> = arrayOf("this", "that", "other", null)
        val expected: Array<String?> = arrayOf("oooo", "iiii", "other", null)

        assertThat(replacer.translate(original)).isEqualTo(expected)
    }

    @Test
    fun `translator with same values should not validate`() {
        assertFailsWith(IllegalArgumentException::class) {
            MetricNameTranslator(mapOf(Pair("this", "that"), Pair("these", "that")))
        }
    }

    @Test
    fun `translator with multiple empty values should validate`() {
        val translator = MetricNameTranslator(mapOf(Pair("this", ""), Pair("these", "")))
        assertThat(translator.translate("this")).isEqualTo("")
        assertThat(translator.translate("these")).isEqualTo("")
    }
}
